kmod/src/forest.c

```diff
@@ -605,14 +662,282 @@ int scoutfs_forest_setup(struct super_block *sb)
 	struct forest_info *finf = SCOUTFS_SB(sb)->forest_info;
 
 	if (finf) {
+		scoutfs_sysfs_destroy_attrs(sb, &finf->ssa);
 		scoutfs_block_put(sb, finf->srch_bl);
+		scoutfs_block_put(sb, finf->oino_bl);
+		vfree(finf->oino_counted_bloom);
 		kfree(finf);
 		sbi->forest_info = NULL;
 	}
 }
+
+/*
+ * Track current open inodes in a counted bloom filter. Save a much smaller
+ * classic BF of open inodes in each transaction, allowing orphan processing to
+ * know when an orphan inode is no longer open on any node.
+ *
+ * Let's call this code: open inode tracking (OINO)
+ */
+
+void scoutfs_forest_oino_inc(struct super_block *sb, u64 ino)
+{
+	__le64 le_ino = cpu_to_le64(ino);
+	DECLARE_FOREST_INFO(sb, finf);
+	u64 hash;
+	int val;
+	int i;
+
+	hash = scoutfs_hash64(&le_ino, sizeof(le_ino));
+	for (i = 0; i < SCOUTFS_FOREST_BLOOM_NRS; i++) {
+		u32 idx = (u32)hash % SCOUTFS_FOREST_BLOOM_BITS;
+		val = atomic_inc_return(&finf->oino_counted_bloom[idx]);
+		WARN_ON_ONCE(val < 0);
+
+		hash >>= SCOUTFS_FOREST_BLOOM_FUNC_BITS;
+	}
+}
+
+void scoutfs_forest_oino_dec(struct super_block *sb, u64 ino)
+{
+	__le64 le_ino = cpu_to_le64(ino);
+	DECLARE_FOREST_INFO(sb, finf);
+	u64 hash;
+	int val;
+	int i;
+
+	hash = scoutfs_hash64(&le_ino, sizeof(le_ino));
+	for (i = 0; i < SCOUTFS_FOREST_BLOOM_NRS; i++) {
+		u32 idx = (u32)hash % SCOUTFS_FOREST_BLOOM_BITS;
+		val = atomic_dec_return(&finf->oino_counted_bloom[idx]);
+		WARN_ON_ONCE(val < 0);
+
+		hash >>= SCOUTFS_FOREST_BLOOM_FUNC_BITS;
+	}
+}
+
+/*
+ * Notify us when a file is orphaned, for debug/test purposes
+ */
+void scoutfs_forest_oino_new_orphan(struct super_block *sb)
+{
+	DECLARE_FOREST_INFO(sb, finf);
+
+	finf->orphan_list_present_hint = true;
+}
+
+/*
+ * At the start of the transaction, alloc and mark dirty a meta block in
+ * anticipation of updating the OINO BF at the end of the transaction.
+ */
+int scoutfs_forest_oino_newtrans_alloc(struct super_block *sb, u64 trans_seq)
+{
+	struct scoutfs_block *new_bl = NULL;
+	struct scoutfs_block_ref *ref;
+	DECLARE_FOREST_INFO(sb, finf);
+	int ret;
+
+	WARN_ON_ONCE(finf->oino_bl);
+
+	ref = &finf->our_log.oino_bloom_ref;
+
+	/* free previous oino block, if any */
+	if (ref->blkno) {
+		ret = scoutfs_free_meta(sb, finf->alloc, finf->wri,
+					le64_to_cpu(ref->blkno));
+		BUG_ON(ret);
+		memset(ref, 0, sizeof(struct scoutfs_block_ref));
+	}
+
+	ret = scoutfs_block_dirty_ref(sb, finf->alloc, finf->wri, ref, SCOUTFS_BLOCK_MAGIC_BLOOM,
+				      &new_bl, 0, NULL);
+	if (ret < 0)
+		goto out;
+
+	finf->oino_bl = new_bl;
+
+out:
+	return ret;
+}
+
+int scoutfs_forest_oino_update(struct super_block *sb)
+{
+	struct scoutfs_bloom_block *bb;
+	DECLARE_FOREST_INFO(sb, finf);
+	u64 last_seq;
+	int ret;
+	int i;
+
+	if (!finf->oino_bl)
+		return 0;
+
+	bb = finf->oino_bl->data;
+
+	ret = scoutfs_client_get_last_seq(sb, &last_seq);
+	if (ret)
+		return ret;
+
+	bb->last_seq = cpu_to_le64(last_seq);
+	finf->our_log.oino_bloom_ref.seq = bb->hdr.seq;
```

Comment:
This shouldn't be overwriting the ref seq.  _dirty_ref already set the ref for it back in newtrans_alloc.

Reply:
ok, omitting that line.